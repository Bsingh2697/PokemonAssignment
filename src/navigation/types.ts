import { ScreenKeys } from "./ScreenKeys";

export type AuthenticationStackParamList = {
    [ScreenKeys.Pokemons] : undefined;
    [ScreenKeys.Details] : undefined;
  };