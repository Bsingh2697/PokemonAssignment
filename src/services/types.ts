
export  type Pokemon = {
    name: string,
    url: string
}

export type Pokemons = {
    count: number,
    next: string,
    previous: string,
    results: Array<Pokemon>
}