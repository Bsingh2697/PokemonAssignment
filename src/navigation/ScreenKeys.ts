export enum ScreenKeys {
    Pokemons = 'Pokemons',
    Details= 'Details'
}