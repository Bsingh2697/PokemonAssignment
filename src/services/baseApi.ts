import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { Pokemon, Pokemons } from "./types";


export const pokemonApi = createApi({
    baseQuery: fetchBaseQuery({
        baseUrl: 'https://pokeapi.co/api/v2/',
    }),
    endpoints:(builder) => ({
        getAllPokemon: builder.query<Pokemons, number>({
            query:(offset : number) => `pokemon/?offset=${offset}&limit=10`,
            serializeQueryArgs: ({ endpointName }) => {
                return endpointName
              },
            merge: (currentCache, newItems) => {
                currentCache.results.push(...newItems.results)
            },
            forceRefetch({ currentArg, previousArg }) {
                return currentArg !== previousArg
              },
        }),
        getPokemonByName: builder.query<Pokemon, string>({
            query: (name) => `pokemon/${name}`,
          }),
    })
})

export const { useGetAllPokemonQuery, useGetPokemonByNameQuery } = pokemonApi

