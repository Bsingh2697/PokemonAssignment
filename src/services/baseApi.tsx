import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { PokemonDetails, Pokemons } from "./types";
import config from 'react-native-config';


export const pokemonApi = createApi({
    baseQuery: fetchBaseQuery({
        baseUrl: config.BASE_URL
    }),
    endpoints:(builder) => ({
        getAllPokemon: builder.query<Pokemons, number>({
            query:(offset : number) => `pokemon/?offset=${offset}&limit=10`,
            serializeQueryArgs: ({ endpointName }) => {
                return endpointName
              },
            merge: (currentCache, newItems) => {
                currentCache.results.push(...newItems.results)
            },
            forceRefetch({ currentArg, previousArg }) {
                return currentArg !== previousArg
              },
        }),
        getPokemonByName: builder.query<PokemonDetails, string>({
            query: (name) => `pokemon/${name}`,
          }),
    })
})

export const { useGetAllPokemonQuery, useGetPokemonByNameQuery } = pokemonApi

