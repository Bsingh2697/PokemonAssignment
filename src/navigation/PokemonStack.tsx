import { createNativeStackNavigator } from "@react-navigation/native-stack";
import { PokemonStackList } from "./types";
import { ScreenKeys } from "./ScreenKeys";
import { PokemonsList } from "../screens/PokemonsList";
import { PokemonDetails } from "../screens/PokemonDetails";


const Stack = createNativeStackNavigator<PokemonStackList>()

export function PokemonStack() {
    return (
        <Stack.Navigator 
        screenOptions={{
            headerShown : false
        }}>
            <Stack.Group>
                <Stack.Screen name={ScreenKeys.Pokemons} component={PokemonsList}/>
                <Stack.Screen name={ScreenKeys.Details} component={PokemonDetails}/>
            </Stack.Group>
        </Stack.Navigator>
    )
}