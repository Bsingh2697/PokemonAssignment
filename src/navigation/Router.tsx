import { NavigationContainer } from "@react-navigation/native";
import { PokemonStack } from "./PokemonStack";

function Router() {
    return (
        <NavigationContainer>
            <PokemonStack/>
        </NavigationContainer>
    )
}

export {Router}