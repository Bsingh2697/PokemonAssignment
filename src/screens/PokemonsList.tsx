import { Text, View } from "native-base"
import { useGetAllPokemonQuery } from "../services/baseApi"
import { Image, StyleSheet, TextInput } from "react-native"
import { StackNavigationProp } from '@react-navigation/stack';
import { FlashList } from "@shopify/flash-list"
import { SkeletonList } from "../components/pokemonCardSkeleton"
import { PokemonCard } from "../components/pokemonCard"
import { useState } from "react"
import { Pokemon } from "../services/types"
import { debounce } from "lodash"
import { ScreenKeys } from "../navigation/ScreenKeys";
import { PokemonStackList } from "../navigation/types";
import { images } from "../assets/images/images";

export interface Props {
    navigation: StackNavigationProp<
      PokemonStackList,
      ScreenKeys.Pokemons
    >;
  }

export const PokemonsList = ({navigation}:Props) => {
    const [page, setPage] = useState(0)
    const { data, error, isLoading, isFetching } = useGetAllPokemonQuery(page)
    const {count, next, results = []} = data || {}
console.log("DATA :L", data)
    const [search, setSearch] = useState('')
    const [searchResult, setSearchResult] = useState<Array<Pokemon>>([])

    const handleSearch = debounce((searchTerm) => {
        const filteredResults = results.filter((pokemon) =>
            pokemon.name.toLowerCase().includes(searchTerm.toLowerCase())
        );
        setSearchResult(filteredResults);
    }, 300);

    const pressPokemon = (name :string) => {
        navigation.navigate(ScreenKeys.Details, {name})
    }

    if(isLoading){
        return (
            <SkeletonList/>
        )
    }

    return(
        <View style={style.listView}>
                <View style={style.headStyle}>
                    <Image source ={images.logo} style={style.logo} testID="pokemon-logo"/>
                    <Text fontFamily="primary" fontWeight="600" fontSize="2xl" marginLeft={4} paddingBottom={8}>POKEMON</Text>
                </View>
            <View marginX="5" marginBottom="4">
                <TextInput onChangeText={(text) => {
                    setSearch(text);
                    handleSearch(text);
                }} placeholder="Find my pokemon!" style={style.textInputStyle}/>
            </View>

            {search.length > 0 ? <FlashList
                    data={searchResult}
                    estimatedItemSize={200}
                    renderItem={({item})=><PokemonCard onPress= {()=>pressPokemon(item.name)} url={item.url} name={item.name}/>}
                /> : 
                <FlashList
                testID="pokemon-list"
                    data={results}
                    estimatedItemSize={200}
                    renderItem={({item})=><PokemonCard onPress= {()=>pressPokemon(item.name)} url={item.url} name={item.name}/>}
                    onEndReached={()=>setPage(page+10)}
                    ListFooterComponent={isFetching ? <SkeletonList showHeader={false}/> : null}
                />
            }
        </View>
    )
}

const style = StyleSheet.create({
    listView:{
        flex:1,
        paddingTop:70,
        backgroundColor:'#ffffff'
    },
    textInputStyle: {borderBottomWidth:1, borderColor:'#ECECEC', fontSize:20},
    logo: {
        width: 100, 
        height: 100, 
      },
      headStyle:{
        flexDirection:'row',
        alignItems:'baseline',
        marginBottom:10
      }
})