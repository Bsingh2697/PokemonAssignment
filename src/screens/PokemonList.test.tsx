import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react-native';
import { PokemonsList } from './PokemonsList';
import { useGetAllPokemonQuery } from '../services/baseApi';
import { StackNavigationProp } from '@react-navigation/stack';
import { PokemonStackList } from '../navigation/types';
import { ScreenKeys } from '../navigation/ScreenKeys';
import { NativeBaseProvider } from 'native-base';

const inset = {
    frame: { x: 0, y: 0, width: 0, height: 0 },
    insets: { top: 0, left: 0, right: 0, bottom: 0 },
  };

const generateStackNavigation = () => ({
    navigate: jest.fn(),
  });

jest.mock('../services/baseApi', () => ({
  useGetAllPokemonQuery: jest.fn(() => ({
    data: {
      count: 10,
      next: null,
      results: [{ name: 'pokemon1', url: 'url1' }, { name: 'pokemon2', url: 'url2' }],
    },
    isLoading: false,
    isFetching: false,
    error: null,
  })),
}));

const navigation = generateStackNavigation as unknown as StackNavigationProp<PokemonStackList, ScreenKeys.Pokemons>

const renderScreen = () => render(<NativeBaseProvider initialWindowMetrics={inset}><PokemonsList navigation={navigation} /></NativeBaseProvider>)

afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

describe('PokemonsList', () => {
  it('renders without crashing', () => {
    renderScreen()
  });
  it('renders header with PokÃ©mon logo and text "POKEMON"', () => {
    renderScreen()
    expect(screen.getByTestId('pokemon-logo')).toBeTruthy();
    expect(screen.getByText('POKEMON'));
  });
  it('renders search input field correctly', () => {
    renderScreen()
    expect(screen.getByPlaceholderText('Find my pokemon!')).toBeTruthy();
  });
  it('matches snapshot', () => {
    renderScreen()
    expect(screen.toJSON()).toMatchSnapshot();
  });
});