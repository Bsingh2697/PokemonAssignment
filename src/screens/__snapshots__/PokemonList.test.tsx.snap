// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonsList matches snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      [
        {},
        {
          "backgroundColor": "#ffffff",
          "flex": 1,
          "paddingTop": 70,
        },
      ]
    }
  >
    <View
      dataSet={{}}
      style={
        [
          {},
          {
            "alignItems": "baseline",
            "flexDirection": "row",
            "marginBottom": 10,
          },
        ]
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/images/logo.png",
          }
        }
        style={
          {
            "height": 100,
            "width": 100,
          }
        }
        testID="pokemon-logo"
      />
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#171717",
            "fontFamily": "primary",
            "fontSize": 24,
            "fontStyle": "normal",
            "fontWeight": "600",
            "letterSpacing": 0,
            "lineHeight": 36,
            "marginLeft": 16,
            "paddingBottom": 32,
            "textDecorationLine": undefined,
          }
        }
      >
        POKEMON
      </Text>
    </View>
    <View
      dataSet={{}}
      style={
        {
          "marginBottom": 16,
          "marginLeft": 20,
          "marginRight": 20,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Find my pokemon!"
        style={
          {
            "borderBottomWidth": 1,
            "borderColor": "#ECECEC",
            "fontSize": 20,
          }
        }
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RCTScrollView
        ListFooterComponent={null}
        applyWindowCorrection={[Function]}
        canChangeSize={true}
        contentContainerStyle={
          {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
        contentHeight={0}
        contentWidth={0}
        data={
          [
            {
              "name": "pokemon1",
              "url": "url1",
            },
            {
              "name": "pokemon2",
              "url": "url2",
            },
          ]
        }
        dataProvider={
          DataProvider {
            "_data": [
              {
                "name": "pokemon1",
                "url": "url1",
              },
              {
                "name": "pokemon2",
                "url": "url2",
              },
            ],
            "_firstIndexToProcess": 0,
            "_hasStableIds": false,
            "_requiresDataChangeHandling": false,
            "_size": 2,
            "getStableId": [Function],
            "rowHasChanged": [Function],
          }
        }
        disableRecycling={false}
        estimatedItemSize={200}
        extendedState={{}}
        externalScrollView={[Function]}
        finalRenderAheadOffset={250}
        forceNonDeterministicRendering={true}
        horizontal={false}
        initialOffset={0}
        initialRenderIndex={0}
        isHorizontal={false}
        layoutProvider={
          GridLayoutProviderWithProps {
            "_acceptableRelayoutDelta": 1,
            "_getHeightOrWidth": [Function],
            "_getLayoutTypeForIndex": [Function],
            "_getSpan": [Function],
            "_hasExpired": false,
            "_maxSpan": 1,
            "_setLayoutForType": [Function],
            "_tempDim": {
              "height": 0,
              "width": 0,
            },
            "averageWindow": AverageWindow {
              "currentAverage": 200,
              "currentCount": 1,
              "inputValues": [
                200,
              ],
              "nextIndex": 1,
            },
            "defaultEstimatedItemSize": 100,
            "layoutObject": {
              "size": undefined,
              "span": undefined,
            },
            "props": {
              "ListFooterComponent": null,
              "data": [
                {
                  "name": "pokemon1",
                  "url": "url1",
                },
                {
                  "name": "pokemon2",
                  "url": "url2",
                },
              ],
              "estimatedItemSize": 200,
              "numColumns": 1,
              "onEndReached": [Function],
              "renderItem": [Function],
              "testID": "pokemon-list",
            },
            "renderWindowInsets": {
              "height": 0,
              "width": 0,
            },
            "shouldRefreshWithAnchoring": true,
          }
        }
        maxRenderAhead={750}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onEndReachedThresholdRelative={0}
        onItemLayout={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onSizeChanged={[Function]}
        onVisibleIndicesChanged={[Function]}
        removeClippedSubviews={false}
        renderAheadOffset={0}
        renderAheadStep={250}
        renderContentContainer={[Function]}
        renderItem={[Function]}
        renderItemContainer={[Function]}
        rowRenderer={[Function]}
        scrollEventThrottle={16}
        scrollThrottle={16}
        scrollViewProps={
          {
            "contentContainerStyle": {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 0,
              "paddingRight": 0,
            },
            "onLayout": [Function],
            "onScrollBeginDrag": [Function],
            "refreshControl": undefined,
            "style": {
              "minHeight": 1,
              "minWidth": 1,
            },
          }
        }
        style={
          {
            "minHeight": 1,
            "minWidth": 1,
          }
        }
        suppressBoundedSizeException={true}
        testID="pokemon-list"
        windowCorrectionConfig={
          {
            "applyToInitialOffset": true,
            "applyToItemScroll": true,
            "value": {
              "endCorrection": 0,
              "startCorrection": 0,
              "windowShift": -0,
            },
          }
        }
      >
        <View>
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <AutoLayoutView
              enableInstrumentation={false}
              horizontal={false}
              onBlankAreaEvent={[Function]}
              onLayout={[Function]}
              renderAheadOffset={0}
              scrollOffset={0}
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
              windowSize={0}
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
